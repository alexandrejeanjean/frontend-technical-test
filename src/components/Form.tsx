import Image from 'next/image';
import { useEffect, useState } from 'react';
import Send from "../assets/send-message.png";
import usePostMessage from '../hooks/usePostMessage';
import { Conversation } from '../types/conversation';
import { getLoggedUserId } from '../utils/getLoggedUserId';

type Props = {
    conversationId: Pick<Conversation, "id">;
    refetchMessages: () => void;
}

const Form = ({ conversationId, refetchMessages }: Props) => {

    if (!conversationId) return null;

    const [message, setMessage] = useState('');
    const authorId = getLoggedUserId();

    const { postMessage, response, loading, error } = usePostMessage({ conversationId: conversationId, authorId: authorId, message: message });


    const resetMessage = () => setMessage('');

    useEffect(() => {
        if (conversationId) resetMessage();
    }, [conversationId])


    const handleChange = (event) => {
        setMessage(event.target.value);
    }

    const handleSubmit = (event) => {
        event.preventDefault();
        postMessage().then(() => {
            resetMessage();
            refetchMessages();
        }).catch(error => {
            console.log("error", error);
        })

    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="messageInput" className="font-bold text-indigo-700">Message</label>
            <div className="relative mt-1.5">
                <label className="sr-only">"Votre message"</label>
                <textarea placeholder=""
                    id="messageInput" name="message" value={message} onChange={handleChange} className="w-full min-h-[50px] pt-3 pl-5 pr-10 bg-gray-100 rounded-md" />
                <button type="submit" className="inline-flex items-center justify-center absolute right-0 top-1/2 w-8 h-8 p-1.5 mr-2.5 bg-primary-blue rounded-full transform -translate-y-1/2">
                    <span className="sr-only">Envoyer le message</span>
                    <Image className="transform -rotate-45" src={Send} alt="logo d'un avion en papier" />
                </button>
            </div>
        </form>
    )
}

export default Form;