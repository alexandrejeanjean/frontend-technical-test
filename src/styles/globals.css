@tailwind base;
@tailwind components;
@tailwind utilities;

.arrow {
  @apply block p-1 border-t-0 border-b-2 border-l-0 border-r-2 border-gray-500 border-solid;
}

.right {
  transform: rotate(-45deg);
  -webkit-transform: rotate(-45deg);
  @apply border-indigo-600;
}

.main-container-height {
  height: calc(100vh - 132px);
}

.list-height {
  height: calc(100% - 50px);
}