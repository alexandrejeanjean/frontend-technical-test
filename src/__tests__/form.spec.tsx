import { expect, jest } from '@jest/globals';
import { shallow } from 'enzyme';
import Form from "../components/Form";


const mockConversation = {
    id: 0,
    lastMessageTimestamp: 1636306788437,
    recipientId: 0,
    recipientNickname: "ThÃ©o Dort",
    senderId: 0,
    senderNickname: "Alexandre Jeanjean"
};

const refetchMessages = jest.fn();



/** @test {Form Component} */
describe('Form Component', () => {

    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Form conversation={mockConversation} refetchMessages={refetchMessages} />);
    });

    it('should render one <form>', () => {
        expect(wrapper.find('form')).toHaveLength(1);
    });

    it('should not render any <button>', () => {
        expect(wrapper.find('button')).toHaveLength(1);
    });

    it('should render 1 <label>', () => {
        expect(wrapper.find('label')).toHaveLength(2);
    });

    it('should render 1 <textarea>', () => {
        expect(wrapper.find('textarea')).toHaveLength(1);
    });

    it('match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
});
