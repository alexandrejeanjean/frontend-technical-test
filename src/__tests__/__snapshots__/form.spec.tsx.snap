// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component match snapshot 1`] = `ShallowWrapper {}`;
